src/gui/qgsattributeform.cpp

```diff
@@ -418,6 +419,66 @@ bool QgsAttributeForm::saveEdits()
   return success;
 }
 
+bool QgsAttributeForm::updateDefaultValues( const int originIdx )
+{
+  if ( !mDefaultValueDependencies.contains( originIdx ) )
+    return false;
+
+  // create updated Feature
+  QgsFeature updatedFeature = QgsFeature( mFeature );
+  if ( mFeature.isValid() || mMode == QgsAttributeEditorContext::AddFeatureMode )
+  {
+    QgsAttributes dst = mFeature.attributes();
+    for ( QgsWidgetWrapper *ww : qgis::as_const( mWidgets ) )
+    {
+      QgsEditorWidgetWrapper *eww = qobject_cast<QgsEditorWidgetWrapper *>( ww );
+      if ( eww )
+      {
+        QVariant dstVar = dst.at( eww->fieldIdx() );
+        QVariant srcVar = eww->value();
+
+        // need to check dstVar.isNull() != srcVar.isNull()
+        // otherwise if dstVar=NULL and scrVar=0, then dstVar = srcVar
+        // be careful- sometimes two null qvariants will be reported as not equal!! (e.g., different types)
+        bool changed = ( dstVar != srcVar && !dstVar.isNull() && !srcVar.isNull() )
```

Comment:
Would `qgsVariantEqual` take care of all that?

Reply:
It would care for the case that when a `NULL` value changes to 0, but for what is mentioned in the comment (I took it from `saveEdit` because there the behavior is the same) `qgsVariantEqual` does not care for the case when a null value does not change, but "two null qvariants will be reported as not equal." This would mean both are `isNull` but the comparison is not true. This could be solved by changing the `qgsVectorEqual` as this:
```
  return lhs.isNull() == rhs.isNull() && lhs == rhs || ( lhs.isNull()  && rhs.isNull() ) ;
```
But I'm not sure about it.

Reply:
I think `lhs.isNull == rhs.isNull ` already takes care of exactly this.

Reply:
But the `&& lhs == rhs` makes it false again. When we pass two `NULL` values. the compared `isNull`s returns true, but are connected with the comparison that could be false (regarding to the comment).

What the `lhs.isNull == rhs.isNull` does is, that it returns false in the end in case we compare `0` with `NULL` what is nice...

Reply:
Just because of the comment or did you experience this?
In the end I think `qgsVariantCompare` was implemented exactly for cases like this, so it would be a pity if it's not usable here.

Reply:
Just because the comment. That's why I'm not sure at all. But it should be same like in `saveEdits` and so I didn't want that I risk issues with this case...

Reply:
Let's see, the check in saveEdits was introduced in 2fddc0079f153c6dd7cb312c8de96548d9812094 (by @nyalldawson)
The method `qgsVariantEqual` later in 90e2c45502457cfc2fa8ea84b011b78d5b4764e1 (by @elpaso)

@nyalldawson / @elpaso, could the above patch be merged into `qgsVariantEqual`?

Reply:
Basically the question is, should we add a testcase for `QVERIFY( QVariant( QVariant::Int ) == QVariant( QVariant::Double ) )`. I'd say yes.

Reply:
+1 to all the above ;)

Reply:
Is it correct that `qgsVariantEqual` should return `true` in case of two `NULL` values from different types, but `false` if one of them is invalid?
Means these test are correct:
```
  QVERIFY( qgsVariantEqual( QVariant( QVariant::Int ), QVariant( QVariant::String ) ) );

  QVERIFY( !qgsVariantEqual( QVariant( QVariant::Int ), QVariant() ) );
```

With that it won't have an issue anymore on the formerly failing test for the `ValueRelationWidget` but what is happening here, is still a little bit weird https://github.com/qgis/QGIS/blob/master/src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp#L236
```
      QVariant v( mComboBox->itemData( i ) );
      if ( qgsVariantEqual( v, value ) )
      {
        idx = i;
        break;
      }
```
There it won't return `true` anymore if `value` is invalid, but still it wouldn't had returned `true` before also if `v` would not be recognized as `NULL` if `mComboBox->itemData( i )` returns `0`. So shouldn't here be set a type with e.g. `toInt()`?